apis/cpi/v1alpha1/cpiconfig_types.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CPIConfigSpec defines the desired state of CPIConfig
+type CPIConfigSpec struct {
+	VSphereCPI VSphereCPI `json:"vsphereCPI"`
+}
+
+type NSXTRoute struct {
+	// NSX-T T0/T1 logical router path
+	// +kubebuilder:validation:Optional
+	RouterPath string `json:"routerPath"`
+}
+
+type NSXT struct {
+	// A flag that enables pod routing
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=false
+	PodRoutingEnabled bool `json:"podRoutingEnabled"`
+
+	// Route configuration for NSXT
+	// +kubebuilder:validation:Optional
+	Routes *NSXTRoute `json:"routes"`
+
+	// A secret reference that contains Nsx-T login credentials to access NSX-T
+	// consists of the field username and password
+	NSXTCredentialsRef v1.SecretReference `json:"nsxTCredentialsRef"`
+
+	// The NSX-T server
+	// +kubebuilder:validation:Optional
+	Host string `json:"host"`
+
+	// InsecureFlag is to be set to true if NSX-T uses self-signed cert
+	// +kubebuilder:validation:Optional
+	InsecureFlag bool `json:"insecureFlag"`
+
+	// RemoteAuth is to be set to true if NSX-T uses remote authentication (authentication done through the vIDM)
+	// +kubebuilder:validation:Optional
+	RemoteAuth bool `json:"remoteAuth"`
+
+	// VMCAccessToken is VMC access token for token based authentication
+	// +kubebuilder:validation:Optional
+	VMCAccessToken string `json:"vmcAccessToken"`
+
+	// VMCAuthHost is VMC verification host for token based authentication
+	// +kubebuilder:validation:Optional
+	VMCAuthHost string `json:"vmcAuthHost"`
+
+	// Client certificate key for NSX-T
+	// +kubebuilder:validation:Optional
+	ClientCertKeyData string `json:"clientCertKeyData"`
+
+	// Client certificate data for NSX-T
+	// +kubebuilder:validation:Optional
+	ClientCertData string `json:"clientCertData"`
+
+	// The certificate authority for the server certificate for locally signed certificates
+	// +kubebuilder:validation:Optional
+	RootCAData string `json:"rootCAData"`
+
+	// The name of secret that stores CPI configuration
+	// +kubebuilder:validation:Optional
+	SecretName string `json:"secretName"`
+
+	// The namespace of secret that stores CPI configuration
+	// +kubebuilder:validation:Optional
+	SecretNamespace string `json:"secretNamespace"`
+}
+
+type ExtraArgs struct {
+	// External arguments for cloud provider
+	// +kubebuilder:validation:Optional
+	TLSCipherSuites string `json:"tls-cipher-suites"`
+}
+
+type NonParavirtualConfig struct {
```

Comment:
NSXT config is the only API spec that should be exposed, rest all should be derived.

Reply:
We discussed this config deriving from control plane machine template if not provided via user input.